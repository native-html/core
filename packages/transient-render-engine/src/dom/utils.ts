import { Text as TextType, Tag as TagType } from 'domelementtype';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === TextType;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === TagType;
}
