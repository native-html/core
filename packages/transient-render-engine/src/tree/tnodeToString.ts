import { TStyles } from '../styles/TStyles';
import { TNode, TNodePrintOptions } from './tree-types';

function printTableStyles(styles: TStyles) {
  const allStyles = {
    ...styles.nativeTextFlow,
    ...styles.nativeBlockFlow,
    ...styles.webTextFlow,
    ...styles.nativeTextRet,
    ...styles.nativeBlockRet
  };
  const entries = Object.entries(allStyles);
  return entries.reduce((prev, [name, val]) => {
    return `${prev}${prev ? ',' : ''} ${name}: ${JSON.stringify(val)}`;
  }, '');
}

function tnodePropertiesString(tnode: TNode, withStyles: boolean) {
  const tagPrint = tnode.tagName ? `tagName="${tnode.tagName}"` : 'anonymous';
  const unregisteredPrint = tnode.isUnregistered ? 'unregistered' : null;
  const indexPrint = `nodeIndex={${tnode.nodeIndex}}`;
  const idPrint = tnode.id ? `id=${tnode.id}` : null;
  const classesPrint = tnode.classes?.length
    ? `classes={[${tnode.classes.join(', ')}]}`
    : null;
  const dataPrint =
    tnode.type === 'text' ? `data=${JSON.stringify(tnode.data)}` : null;
  const anchorPrint =
    typeof tnode.attributes.href === 'string'
      ? `href=${JSON.stringify(tnode.attributes.href)}`
      : null;
  const srcPrint =
    typeof tnode.attributes.src === 'string'
      ? `src=${JSON.stringify(tnode.attributes.src)}`
      : null;
  const pstyles = withStyles ? printTableStyles(tnode.styles) : null;
  const stylesPrint = pstyles ? `styles={{${pstyles} }}` : null;
  const detailsPrint = [
    tagPrint,
    unregisteredPrint,
    indexPrint,
    idPrint,
    classesPrint,
    dataPrint,
    anchorPrint,
    srcPrint,
    stylesPrint
  ]
    .filter((p) => p !== null)
    .join(' ');
  return `<${tnode.displayName} ${detailsPrint}`;
}

interface TNodePrintState {
  parentLeftPrefix: string;
  isChild: boolean;
  isLast: boolean;
}

export default function serializeTNode(
  tnode: TNode,
  params: Partial<TNodePrintState & TNodePrintOptions> = {}
): string {
  const {
    parentLeftPrefix = '',
    isChild = false,
    isLast = false,
    withStyles = true,
    withNodeIndex = true,
  } = params;
  const prefix = isChild ? '  ' : '';
  const totalPrefixLeft = parentLeftPrefix + prefix;
  const childrenPrint = tnode.children
    .map((c, i) =>
      serializeTNode(c, {
        parentLeftPrefix: parentLeftPrefix + ' '.padStart(prefix.length, ' '),
        isChild: true,
        isLast: i === tnode.children.length - 1,
        withStyles,
        withNodeIndex
      })
    )
    .join('');
  return `${totalPrefixLeft}${tnodePropertiesString(tnode, withStyles)}${
    childrenPrint
      ? `>\n${childrenPrint}\n${totalPrefixLeft}</${tnode.displayName}>`
      : ' />'
  }${isLast ? '' : '\n'}`;
}
