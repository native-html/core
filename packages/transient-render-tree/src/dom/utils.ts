import { ElementType } from 'htmlparser2';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === ElementType.Text;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === ElementType.Tag;
}
