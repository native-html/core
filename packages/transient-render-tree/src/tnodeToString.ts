import { TNode } from './tree/TNode';
import { TText } from './tree/TText';

function tnodePropertiesString(tnode: TNode) {
  const tagPrint = tnode.tagName ? `tag=${tnode.tagName}` : 'anonymous';
  const idPrint = tnode.id ? `id=${tnode.id}` : null;
  const classesPrint = tnode.className ? `classes=${tnode.className}` : null;
  const dataPrint =
    tnode instanceof TText
      ? `data[${
          tnode.data.length > 24
            ? JSON.stringify(tnode.data.substring(0, 24)) + '…'
            : JSON.stringify(tnode.data)
        }]`
      : null;
  // @ts-ignore
  const anchorPrint = tnode.isAnchor ? `anchor[${tnode.href}]` : null;
  const detailsPrint = [tagPrint, idPrint, classesPrint, dataPrint, anchorPrint]
    .filter((p) => p !== null)
    .join(',');
  return `•${tnode.displayName}(${detailsPrint})`;
}

interface TNodePrintState {
  parentLeftPrefix: string;
  isChild: boolean;
  isLast: boolean;
}

export default function tnodeToString(
  tnode: TNode,
  {
    parentLeftPrefix = '',
    isChild = false,
    isLast = false
  }: Partial<TNodePrintState> = {}
) {
  const prefix = isChild ? (isLast ? '└─' : '├─') : '';
  const totalPrefixLeft = parentLeftPrefix + prefix;
  const childrenPrint: string = tnode.children
    .map((c, i) =>
      tnodeToString(c, {
        parentLeftPrefix:
          parentLeftPrefix +
          ' '.padStart(prefix.length, isChild && isLast ? ' ' : '│'),
        isChild: true,
        isLast: i === tnode.children.length - 1
      })
    )
    .join('');
  return `${totalPrefixLeft}${tnodePropertiesString(tnode)}\n${childrenPrint}`;
}
