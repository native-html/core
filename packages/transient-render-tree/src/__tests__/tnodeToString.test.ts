import tnodeToString from '../tnodeToString';
import { translateTreeTest } from '../flow/__tests__/utils';

describe('tnodeToString', () => {
  it('should handle deeply nested tags', () => {
    const result = tnodeToString(
      translateTreeTest(
        '<div><div><span>A<em>a</em></span><span>B</span></div><span>C<strong>D</strong></span></div>'
      )
    );
    expect(result).toMatchSnapshot();
  });
});
