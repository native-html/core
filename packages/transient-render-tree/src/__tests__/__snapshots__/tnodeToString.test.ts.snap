// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tnodeToString should handle deeply nested tags 1`] = `
"•TBlock(tag=div)
 ├─•TBlock(tag=div)
 │ ├─•TPhrasing(tag=span)
 │ │ ├─•TText(anonymous,data=\\"A\\")
 │ │ └─•TText(tag=em,data=\\"a\\")
 │ └─•TText(tag=span,data=\\"B\\")
 └─•TPhrasing(tag=span)
   ├─•TText(anonymous,data=\\"C\\")
   └─•TText(tag=strong,data=\\"D\\")
"
`;
